### PR TITLE
Ensure Tag(Keys|Values) APIs never return (nil, nil)

### PR DESCRIPTION
**Pull request by stuartcarnie**
Formalized this post condition in the documentation and added additional
unit tests.

Added a nil guard and unit test to WriteStringIterator.

### CODE REVIEW

File: tsdb/tsm1/engine_schema_test.go [hunk 1]
```diff
@@ -238,6 +238,23 @@ memB,host=EB,os=macOS value=1.3 201`)
 			exp:      []string{"0B", "AB", "BB", "CB", "DB", "EB"},
 			expStats: cursors.CursorStats{ScannedValues: 3, ScannedBytes: 24},
 		},
+
+		// ***********************
+		// * other scenarios
+		// ***********************
+		{
+			// ensure StringIterator is never nil
+			name: "predicate/no candidate series",
```
Comment (jlapacik) [inline tsdb/tsm1/engine_schema_test.go]: I'm assuming this test case would panic previously?

File: storage/reads/string_iterator_writer.go [hunk 2]
```diff
@@ -38,6 +38,10 @@ func (w *StringIteratorWriter) WrittenN() int {
 }
 
 func (w *StringIteratorWriter) WriteStringIterator(si cursors.StringIterator) error {
+	if si == nil {
```
Comment (jlapacik) [inline storage/reads/string_iterator_writer.go]: This should never be true, right? Just a precaution?

File: tsdb/tsm1/engine_schema_test.go (hunk 1 above)
Comment (stuartcarnie) [reply tsdb/tsm1/engine_schema_test.go]: correct – no more panic

File: storage/reads/string_iterator_writer.go (hunk 2 above)
Comment (stuartcarnie) [reply storage/reads/string_iterator_writer.go]: right - only precautionary

**Pull request merged**